warehouse/warehouse.go

```diff
@@ -288,22 +293,54 @@ func (wh *HandleT) backendConfigSubscriber(ctx context.Context) {
 		wh.workspaceBySourceIDsLock.Lock()
 		wh.workspaceBySourceIDs = map[string]string{}
 
-		pkgLogger.Infof(`Received updated workspace config`)
+		controlPlaneAPI := cpclient.NewInternalClientWithCache(
+			wh.configBackendURL,
+			wh.internalAuth)
+
+		pkgLogger.Info(`Received updated workspace config`)
 		for workspaceID, wConfig := range config {
+
 			for _, source := range wConfig.Sources {
 				if _, ok := sourceIDsByWorkspace[workspaceID]; !ok {
 					sourceIDsByWorkspace[workspaceID] = []string{}
 				}
+
 				sourceIDsByWorkspace[workspaceID] = append(sourceIDsByWorkspace[workspaceID], source.ID)
 				wh.workspaceBySourceIDs[source.ID] = workspaceID
 
 				if len(source.Destinations) == 0 {
 					continue
 				}
-				for _, destination := range source.Destinations {
+
+				for idx, destination := range source.Destinations {
+
 					if destination.DestinationDefinition.Name != wh.destType {
 						continue
 					}
+
+					if _, ok := destination.Config["useSSH"]; ok {
+						source.Destinations[idx].Config["tunnellingType"] = "ssh_forward"
+					}
+
+					// When tunnel object is set from upstream, we need to fetch
+					// the keypair associated with the destinationId.
+					// Only assume valid tunnellingType to be there.
+					if _, ok := destination.Config["tunnellingType"]; ok {
+
+						keypair, err := controlPlaneAPI.GetDestinationSSHKeys(ctx, destination.ID)
+						if err != nil {
+							pkgLogger.Warnf("fetching ssh keypair for destination: %s err: %w", destination.ID, err)
+						}
+
+						if keypair == nil {
+							pkgLogger.Warnf("unable to locate keypair for destination: %s with valid tunnel info", destination.ID)
+						} else {
+							source.Destinations[idx].Config["privateKey"] = keypair.PrivateKey
+							source.Destinations[idx].Config["publicKey"] = keypair.PublicKey
+						}
+
+					}
```

Comment:
I would use a middleware approach for intercepting and injecting tunnel config.

Similar to how `tenantManager.WatchConfig` works. This will help with readability and testability of the newly added code, while keeping this section clean.